### DISCUSSION TITLE
zfs-windows-2.2.6-rc6 & zfs-windows-2.2.6-rc7

### DISCUSSION THREAD

**Discussion by lundman** [openzfsonwindows/openzfs#402] [Development]
I have posted a new build for rc6 / rc7.

I believe I have fixed setting driveletter, mounting of snapshots, some unmounting issues.

So if you can, please check your issues again, even if I didn't list them, and remind me of any issues that are still present.

**Comment by leonik3** [comment]
I installed this version. Unfortunately it destroyed my partitions on two other drives.

1. I launched:
zpool create -o autoexpand=on -o autotrim=on -O dedup=on -O casesensitivity=insensitive -O compression=on -O atime=off -o ashift=12 zzz raidz PHYSICALDRIVE4 PHYSICALDRIVE9 PHYSICALDRIVE10

2. Windows got a blue screen.

3. Windows didn't start anymore

4. On another computer I saw that zfs partitions were saved on two completely different disks. The first was the Windows system disk on which I was running the command. The second was a disk with some Ubuntu. I have other disks, but they are as connected dynamic disks in Windows - they were not destroyed.

5. I recovered the Windows partition and Windows system using the following methods:
a) Testdisk-7.3-WIP and "testdisk_win.exe" - all partitions was recovered
b) Windows installation disk and repair mode in cmd and chkdsk /f
c) Windows installation disk and repair mode in cmd and:
- bootrec /fixmbr
- bootrec /fixboot
- bootrec /rebuildbcd
d) old version of https://sourceforge.net/projects/boot-repair-cd/
e) Once more Windows installation disk and repair mode in cmd and:
- bootrec /fixmbr
- bootrec /fixboot
- bootrec /rebuildbcd

**Comment by BellCubeDev** [comment]
On my machine, I can confirm that exporting a pool now allows it to be automatically mounted in my Linux Mint dual-boot setup (something I wasn't sure was possible before). Whatever you did with the export and unmount code, it appears to work more correctly than before.

**Comment by guenther-alka** [comment]
I was abroad for some time but now found the time to update my laptop with 2.2.3rc4 to 2.2.6 rc7.
Update ended with a BSOD, after reboot zpool import command does not work.

I deinstalled rc7 and reinstalled: ok

I had created a pool on some vhdx virtual harddisks in rc4 and tried to import after mounting them. 
This ends always in a BSOD (openzfs.sys, system thread exception not handled)

see https://drive.google.com/file/d/1WuVJZXNpXdx5Eg5iM1iGogRXZoYmHVLv/view?usp=sharing 

I created a new virtual harddisk and tried to create a new pool on it: same BSOD

**Reply by lundman** [reply]
OK so you get BSODs with rc7?

**Comment by guenther-alka** [comment]
zpool version
zfswin-2.2.6rc7
zfs-kmod-zfswin-2.2.6rc7

BSOD when creating or importing a pool on virtual harddisks .vhdx (win 11 pro 23H2)

**Comment by guenther-alka** [comment]
BSOD also happens with iSCSI targets or a local USB disk
Is this still related to the Aomei driver?

**Comment by guenther-alka** [comment]
I have deinstalled Aomei and it works now,
so the problem is related to rhe known incompatibility with Aomei

btw
When I try to access a snap in explorer ex d:\data\.zfs\snapshot\snapname I get "access denied"

![image](https://github.com/user-attachments/assets/81c7be75-075d-4cc9-ac14-a4357dbe1f54)

when I try again I get a file icon instead a folder icon
![image](https://github.com/user-attachments/assets/827bdd70-e28a-4f37-9e42-9410f861e18c)

I then tried to set a driveletter k to tank/data
I could not open the snapshot folder (endless "beachball", not visible in screenshot) 

![image](https://github.com/user-attachments/assets/8f1a7229-c1fd-4a56-8be5-d1427bbdd15a)

**Reply by sskras** [reply]
Since [2.2.3-rc6](https://github.com/openzfsonwindows/openzfs/compare/4be6e28%5E...zfswin-2.2.3rc6) seems to include a work around for `ambakdrv.sys`, I suspect the root cause to still be coming from that driver.

**Reply by lundman** [reply]
So you crash with `ambakdrv.sys` with default settings in the tunable? It is supposed to skip loading zvol support to work still, and only if you set it to `2` will it force load zvol.  I might have to go back and check that again then, be nice if it worked by default.

**Comment by guenther-alka** [comment]
windows_zvol_enabled is set to 0

**Comment by guenther-alka** [comment]
The easiest and fastest method to replicate filesystems between any ZFS OS is netcat.
This works perfect between BSD, Linux, OSX and Solaris/Illumos
http://www.unixsamurai.com/zfs-send-and-receive-with-netcat/

I tried again with rc7 without success from/to Windows with nc64 and nmap netcat (failed to read from stream)
My preferred netcat on Windows would be the free single binary nc64.exe v1.12 (nc32 is reported as hackertool in defender)
https://eternallybored.org/misc/netcat/

It would be great if netcat replication works with ZFS on Windows

**Reply by lundman** [reply]
great I'll check that tomorrow, I was looking for something to do

**Reply by lundman** [reply]
Using CMD (Administrator) has it almost working: (CMD as pipes do not work yet when using MSYS or PS)
```
CMD1> nc64.exe -l -p 40000 | out\build\x64-Debug\cmd\zfs\zfs recv BOOM/dataset

CMD2> out\build\x64-Debug\cmd\zfs\zfs send BOOM@send | nc64.exe 127.0.0.1 40000
09:49:00    312B   BOOM@send
09:49:02   4.91M   BOOM@send
09:49:02   5.04M   BOOM@send
09:49:02   5.16M   BOOM@send
^C
```

Biggest issue appears to be that it does not terminate the connection when the send is complete. 

(Additionally, I did not use -v for verbose send, and yet it is on)

**Reply by lundman** [reply]
OK, so it is a netcat issue
```
type README.md | nc64.exe 127.0.0.1 40000
```

Will send the file, but not close the connection when `type` exits. We can probably patch netcat to handle the Windows equivalent of `SIGCHLD` - I mean, if we find time.


However, changing Windows sending side (only) to use `ncat` from `nmap` (https://nmap.org/dist/ncat-portable-5.59BETA1.zip) works without changes:

`out\build\x64-Debug\cmd\zfs\zfs send BOOM@send | ncat.exe --send-only 127.0.0.1 40000`

`zfs send` terminates, `ncat` quits, connection is dropped and the receiving `netcat` also closes.

**Comment by guenther-alka** [comment]
Thanks, I tried the single binary nmap netcat and nc64

On Windows use cmd as powershell has problem with pipe commands
Replication works but does not terminate

```
nmap netcat

##############
proxmox -> win
##############

must use cmd, ps not working with pipe !!

PS C:\Users\me> cmd
Microsoft Windows [Version 10.0.22631.4317]
(c) Microsoft Corporation. Alle Rechte vorbehalten.

C:\Users\me>C:\xampp\web-gui\ncat\ncat.exe -v -l -p 53087 | zfs receive -v tank/pve
Ncat: Version 5.59BETA1 ( http://nmap.org/ncat )
Ncat: Listening on 0.0.0.0:53087
Ncat: Connection from 192.168.2.71:59036.
receiving full stream of tank/data@1 into tank/pve@1
received 43.9K stream in 1.85 seconds (23.7K/sec)

then start sender

root@pve:~# zfs send tank/data@1 | pv | nc 192.168.2.65 53087
44.2KiB 0:00:00 [1.99MiB/s] [<=>

Problem:
tank/pve and tank/pve@1 created but send and receive does not terminate

C:\Users\me> zfs list -t snap,filesystem
NAME         USED  AVAIL  REFER  MOUNTPOINT
tank         334K  13.6G    38K  /tank
tank/pve    46.5K  13.6G  24.5K  /tank/pve
tank/pve@1    22K      -    24K  -


##############
win -> proxmox
##############

####
send
####

root@pve:~# nc -l -p 53087 | zfs receive -v tank/win
receiving full stream of tank/data@1 into tank/win@1
received 47.1K stream in 1.71 seconds (27.6K/sec)

C:\Users\me>zfs send tank/data@1 | C:\xampp\web-gui\ncat\ncat.exe -v 192.168.2.71 53087
10:20:24    312B   taNcat: nk/data@1
Version 5.59BETA1 ( http://nmap.org/ncat )
10:20:24    312B   tank/data@1
10:20:24   4.09K   tank/data@1
.. repeats this line a hundredfold
Ncat: Connected to 192.168.2.71:53087.
10:20:24   4.09K   tank/data@1
.. repeats this line a hundredfold
10:20:24   13.2K   tank/data@1
..repeats this line a hundredfold
10:20:24   30.2K   tank/data@1
..repeats this line a hundredfold
10:20:25   46.5K   tank/data@1
..repeats this line a hundredfold

then waiting, does not terminate

#######
receive
#######

root@pve:~# nc -l -p 53087 | zfs receive -v tank/win
receiving full stream of tank/data@1 into tank/win@1
received 47.1K stream in 1.71 seconds (27.6K/sec)

then waiting, does not terminate


root@pve:~# zfs list -t snapshot,filesystem | grep tank/win
tank/win                                              133K  1.40G   133K  /tank/win
tank/win@1                                              0B      -   133K  -
root@pve:~#


```

ok, so replication works but does not terminate

**Reply by lundman** [reply]
Use `ncat` for the `send` process. `nc` worked on `recv`. Don't forget the `--send-only`

**Comment by guenther-alka** [comment]
ok
nmap netcat replicates and terminated proper with --send-only and zfs send

```
C:\Users\me>zfs send tank/data@1 | C:\xampp\web-gui\ncat\ncat.exe -v --send-only 192.168.2.71 53087
11:25:11    312B   tank/data@1
..
Ncat: 48520 bytes sent, 0 bytes received in 0.11 seconds.
C:\Users\me>
```

As it is not helpful to use a different netcat for send and receive, nmap netcat is it then at the moment

receive works but does not terminate (tried also with --recv-only
```
C:\Users\me>C:\xampp\web-gui\ncat\ncat.exe -v -l --recv-only -p 53087 | zfs receive -v tank/pve4
Ncat: Version 5.59BETA1 ( http://nmap.org/ncat )
Ncat: Listening on 0.0.0.0:53087
Ncat: Connection from 192.168.2.71:60794.
receiving full stream of tank/data@1 into tank/pve4@1
received 43.9K stream in 1.48 seconds (29.8K/sec)
(does not terminate but tank/pve4 and snap was created)
```

**Reply by lundman** [reply]
I only testing with nc.exe for recv, I did have to wait some 10s after send finished before recv quit, I guess the buffering is different.

It's quite fine to mix nc and ncat. It plain text connections after all

**Comment by guenther-alka** [comment]
receive not terminated (now after 10min) with ncat

btw
I usually include one dedicated netcat even on Linux or BSD to make it copy and run and to avoid different command options or different versions. Nc is complicated enough due different needed options on BSD and Linux. 

Nmap ncat cannot be redistributed but must be downloaded by user but that can be handled, nc64 can be redistributed.

**Comment by lundman** [comment]
Sorry for the pause, the EV certificate is expiring March 8, so going through the renewal process - which takes about a week for some reason. (I can pay $699 to expedite it though...)

**Reply by lundman** [reply]
OK new certificate installed and we are ready to resume. What are the bug priorities :)

**Comment by guenther-alka** [comment]
Depends whether problems depend on each other but problem with large files is annoying.
(zvol destroy not possible in latest but zvol not as important as ZFS filesystem handling)

**Reply by lundman** [reply]
Teach me about this large file problem?

**Comment by guenther-alka** [comment]
when I copy a large file ex win64.iso from c: to a ZFS volume test/data
copy stalles after around 1GB

When I cancel copy, file is visible at ZFS. When I delete it remains visible on ZFS after a refresh in Explorer

zpool export test last a few minutes after this

```
zpool export test
zunmount(test/data/2,D:/data/2/) running
zunmount(test/data/2,D:/data/2/) returns 0
zunmount(test/data/1,D:/data/1/) running
zunmount(test/data/1,D:/data/1/) returns 0
zunmount(test/data,D:/data/) running
zunmount(test/data,D:/data/) returns 0
zunmount(test,D:/) running
zunmount(test,D:/) returns 0
PS C:\Users\me> 
```

![image](https://github.com/user-attachments/assets/bb4654f0-29a1-4c55-816b-bfe48f478a8d)

**Reply by lundman** [reply]
Ah ok thanks. The "cancel" operation has not been implemented, so dont change your mind and cancel things. :)

**Comment by guenther-alka** [comment]
btw
https://github.com/openzfsonwindows/openzfs/releases/download/zfswin-2.3.0rc6/OpenZFSOnWindows-debug-.99-3528-gf8b636490.exe fixes the zfs destroy zvol problem

Is it intended that this thread is about 2.2.6rc?

**Reply by lundman** [reply]
Just used it for EV PSA, not zvol etc?
